### PR TITLE
amp-bind: Refactor attribute mutation flow

### PR DESCRIPTION
**Pull request by choumx**
Partial for #6199.

- Changed attribute mutation to use `Resources#mutateElement` instead of `vsync` 
- Use a single callback per element with N mutations rather than N callbacks
- Avoid overwriting AMP CSS classes when applying `[class]` binding
- Pass additional data to `BindEvaluator` for performing runtime validation (to be added later)
- Convert new attribute values to `String` rather than restricting by type

Design details here: https://docs.google.com/a/google.com/document/d/1oQeCELbc_jjwH7MiP8kohYlHcP_XSXlPwhqPQ-KgVhI/pub#h.r5ma7fwbacw4

### CODE REVIEW

**Comment by choumx** [comment]
/to @dvoytenko /to @jridgewell /cc @cramforce 

PTAL

File: extensions/amp-bind/0.1/bind-impl.js [hunk 1]
```diff
@@ -100,9 +94,9 @@ export class Bind {
     };
 
     this.ampdoc.whenBodyAvailable().then(body => {
-      const {bindings, expressions} = this.scanForBindings_(body);
-      this.bindings_ = bindings;
-      this.evaluator_ = new BindEvaluator(expressions);
+      const {boundElements, evaluatees} = this.scanForBindings_(body);
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: Is it not too early to scan for bindings?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 2]
```diff
@@ -128,235 +123,268 @@ export class Bind {
...
     // TODO(choumx): Chunk if taking too long in a single frame.
 
-    const bindings = [];
-    const expressions = [];
+    /** @type {!Array<BoundElementDef>} */
+    const boundElements = [];
+    /** @type {!Array<./bind-evaluator.EvaluateeDef>} */
+    const evaluatees = [];
+
     const elements = body.getElementsByTagName('*');
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: This is a big scan. Should we use treewalker instead?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 3]
```diff
@@ -128,235 +123,268 @@ export class Bind {
...
     const name = attribute.name;
-    if (name.length > 2) {
-      if (name[0] === '[' && name[name.length - 1] === ']') {
-        return {
-          property: name.substr(1, name.length - 2),
-          expression: attribute.value,
-          element,
-        };
-      }
+    if (name.length > 2 && name[0] === '[' && name[name.length - 1] === ']') {
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: Did we confirm if these attribute names valid HTML? I've been trying to set it via `Element.setAttribute` with no luck.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 2 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: I'm actually planning on taking a closer look at performance of this (per the TODO above) and other initialization tasks like the parsing of expressions. Justin also suggested using attribute existence query selector. I'll check out Treewalker too, thanks for the tip!

File: extensions/amp-bind/0.1/bind-impl.js (hunk 3 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: It is technically valid per [W3 spec](https://html.spec.whatwg.org/multipage/syntax.html#attributes-2), though it seems some browsers like Chrome are more strict in `setAttribute`.

> Attribute names must consist of one or more characters other than the space characters, U+0000 NULL, U+0022 QUOTATION MARK ("), U+0027 APOSTROPHE ('), U+003E GREATER-THAN SIGN (>), U+002F SOLIDUS (/), and U+003D EQUALS SIGN (=) characters, the control characters, and any characters that are not defined by Unicode.

Angular also uses [similar syntax for templates](https://angular.io/docs/ts/latest/guide/template-syntax.html#!#binding-syntax).

File: extensions/amp-bind/0.1/bind-impl.js (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Do you think `whenFirstVisible` is more appropriate? Agree that we should make sure it doesn't affect time-to-first-paint.

Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: I think we need `ampdoc.whenReady()` here. Basically, `whenBodyAvailable` means when document's opening `<body>` has been parsed. `whenReady` means that final `</html>` has been parsed.

That being said, we need to ensure that we prioritize it properly - it's a lot of work to do CPU-wise and it'd be unideal if we fight over resources with initial rendering. Not sure how we can do this yet. I think it's ok to add a `TODO` and file an issue for now.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 3 above)
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: But did we look in all supported browsers if that works?

File: extensions/amp-bind/0.1/bind-impl.js (hunk 1 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Thanks for the context. Changed to `whenReady`.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 3 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Yep, `test-bind-impl.js` uses the bracket syntax for unit tests and is running on Travis.

**Comment by choumx** [comment]
Filed an issue for init performance with some of my profiling results in #6910.

File: builtins/amp-img.js [hunk 4]
```diff
@@ -47,14 +47,21 @@ export class AmpImg extends BaseElement {
...
-      this.updateImageSrc_();
-    } else if (this.img_ && ATTRIBUTES_TO_PROPAGATE.indexOf(name) >= 0) {
-      this.propagateAttributes(name, this.img_,
-          /* opt_removeMissingAttrs */ true);
-    }
+  mutatedAttributesCallback(mutations) {
+    mutations.forEach(mutation => {
+      const name = mutation.name;
+
+      if (name === 'src') {
```
Comment (dvoytenko) [inline builtins/amp-img.js]: Just `==`

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 5]
```diff
@@ -17,34 +17,52 @@
 import {BindExpression} from './bind-expression';
 
 /**
+ * @typedef {{
+ *   tagName: string,
+ *   property: string,
+ *   expressionString: string,
+ *   expression: (BindExpression|undefined),
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Looks like it's actually `!BindExpression|undefined`, right?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 6]
```diff
@@ -128,235 +124,268 @@ export class Bind {
...
+        });
+
+        if (width !== undefined || height !== undefined) {
+          // TODO(choumx): Add new Resources method for adding change-size
+          // request without scheduling vsync pass since `mutateElement()`
+          // will schedule a pass after a short delay anyways.
+          this.resources_./*OK*/changeSize(element, height, width);
+        }
+
+        if (typeof element.mutatedAttributesCallback === 'function') {
```
Comment (dvoytenko) [inline extensions/amp-bind/0.1/bind-impl.js]: We can defer to the followup PRs, but some significant questions here are:

1. What if the target is a non-AMP element? E.g. consider `<div hidden>` attribute.
2. What if the target AMP element has not been upgraded yet?

File: builtins/amp-img.js (hunk 4 above)
Comment (choumx) [inline builtins/amp-img.js]: Done.

File: extensions/amp-bind/0.1/bind-evaluator.js (hunk 5 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 6 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: 1. Non-AMP elements don't need additional handling via `mutatedAttributesCallback`. Is there an issue here? 
2. The non-upgraded AMP element will have its attribute set above. Once upgraded and built, it should work normally (as you mentioned in email).

Though I'll keep these in mind when I further refactor this to use a middleware layer per our email discussion.

File: builtins/amp-img.js [hunk 7]
```diff
@@ -47,14 +47,21 @@ export class AmpImg extends BaseElement {
...
-  attributeChangedCallback(name, unusedOldValue, unusedNewValue) {
-    if (name === 'src') {
-      this.srcset_ = srcsetFromElement(this.element);
-      this.updateImageSrc_();
-    } else if (this.img_ && ATTRIBUTES_TO_PROPAGATE.indexOf(name) >= 0) {
-      this.propagateAttributes(name, this.img_,
-          /* opt_removeMissingAttrs */ true);
-    }
+  mutatedAttributesCallback(mutations) {
+    mutations.forEach(mutation => {
```
Comment (jridgewell) [inline builtins/amp-img.js]: Is the `#forEach` necessary?

File: extensions/amp-bind/0.1/bind-evaluator.js [hunk 8]
```diff
@@ -17,34 +17,52 @@
...
+  constructor(evaluatees) {
+    /** @const {!Array<EvaluateeDef>} */
+    this.evaluatees_ = [];
+
+    // TODO(choumx): Add expression result validation to this class.
+    evaluatees.forEach(e => {
+      this.evaluatees_.push({
+        tagName: e.tagName,
+        property: e.property,
+        expressionString: e.expressionString,
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Nit: should this be on `BindExpression`?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 9]
```diff
@@ -128,235 +124,268 @@ export class Bind {
...
-
     this.evaluatePromise_ = this.evaluator_.evaluate(this.scope_);
     this.evaluatePromise_.then(results => {
-      /** {!VsyncTaskSpecDef} */
-      const task = {measure: this.boundMeasure_, mutate: this.boundMutate_};
-      this.vsync_.run(task, {
-        results,
-        verifyOnly: !!opt_verifyOnly,
-      });
+      if (!!opt_verifyOnly) {
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: Why `!!`?

File: extensions/amp-bind/0.1/bind-impl.js [hunk 10]
```diff
@@ -128,235 +124,268 @@ export class Bind {
...
    * @private
    */
-  measure_(unusedState) {
-    // TODO(choumx): Validate here or in applyBinding_()?
+  verify_(results) {
+    this.boundElements_.forEach(boundElement => {
+      const {element, bindings} = boundElement;
+
+      bindings.forEach(binding => {
+        const newValue = results[binding.expression];
```
Comment (jridgewell) [inline extensions/amp-bind/0.1/bind-impl.js]: `expressionString`?

File: builtins/amp-img.js (hunk 7 above)
Comment (choumx) [inline builtins/amp-img.js]: No, but is it ever strictly necessary? :) 

Changed to `for` loop.

File: extensions/amp-bind/0.1/bind-evaluator.js (hunk 8 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-evaluator.js]: Fair, done.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 9 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: A vestige of the old code. Removed.

File: extensions/amp-bind/0.1/bind-impl.js (hunk 10 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Done.

File: builtins/amp-img.js (hunk 7 above)
Comment (jridgewell) [inline builtins/amp-img.js]: No, I mean, why does it matter what attributes changed? Let's just propagate. We could do an `#find` if creating a new srcset is expensive.

**Comment by dvoytenko** [comment]
LGTM on my side.

File: builtins/amp-img.js [hunk 11]
```diff
@@ -47,13 +47,20 @@ export class AmpImg extends BaseElement {
...
   /** @override */
-  attributeChangedCallback(name, unusedOldValue, unusedNewValue) {
-    if (name === 'src') {
-      this.srcset_ = srcsetFromElement(this.element);
-      this.updateImageSrc_();
-    } else if (this.img_ && ATTRIBUTES_TO_PROPAGATE.indexOf(name) >= 0) {
-      this.propagateAttributes(name, this.img_,
-          /* opt_removeMissingAttrs */ true);
+  mutatedAttributesCallback(mutations) {
+    for (let i = 0; i < mutations.length; i++) {
```
Comment (jridgewell) [inline builtins/amp-img.js]: Re: https://github.com/ampproject/amphtml/pull/6779#discussion_r95037665

No, I mean, why does it matter what attributes changed? Let's just propagate. We could do an #find if creating a new srcset is expensive.

Comment (choumx) [inline builtins/amp-img.js]: Changed to remove need for iteration as discussed.

File: builtins/amp-img.js [hunk 12]
```diff
@@ -47,13 +47,20 @@ export class AmpImg extends BaseElement {
...
-    } else if (this.img_ && ATTRIBUTES_TO_PROPAGATE.indexOf(name) >= 0) {
-      this.propagateAttributes(name, this.img_,
-          /* opt_removeMissingAttrs */ true);
+      // This element may not have been laid out yet.
+      if (this.img_) {
+        this.updateImageSrc_();
+      }
+    }
+
+    if (this.img_) {
```
Comment (jridgewell) [inline builtins/amp-img.js]: Nit: We can create one `if (img)`, and run both cases inside it.

File: builtins/amp-img.js [hunk 13]
```diff
@@ -47,13 +47,20 @@ export class AmpImg extends BaseElement {
...
+      const attrs = ATTRIBUTES_TO_PROPAGATE.filter(
```
Comment (cramforce) [inline builtins/amp-img.js]: This should maybe become a filter at some point.

File: extensions/amp-bind/0.1/bind-impl.js [hunk 14]
```diff
@@ -128,235 +124,267 @@ export class Bind {
...
+        }
+        break;
+
+      default:
+        const oldValue = element.getAttribute(property);
+
+        let attributeChanged = false;
+        if (typeof newValue === 'boolean') {
+          if (newValue && oldValue !== '') {
+            element.setAttribute(property, '');
```
Comment (cramforce) [inline extensions/amp-bind/0.1/bind-impl.js]: Does this happen in the same instance as the `mutatedAttributesCallback`?

File: builtins/amp-img.js (hunk 13 above)
Comment (choumx) [inline builtins/amp-img.js]: Not sure what you mean by "become a filter". A separate function?

File: builtins/amp-img.js (hunk 12 above)
Comment (choumx) [inline builtins/amp-img.js]: We always want to update `this.srcset_` though, how would that work without duplicating the first conditional?

File: extensions/amp-bind/0.1/bind-impl.js (hunk 14 above)
Comment (choumx) [inline extensions/amp-bind/0.1/bind-impl.js]: Yes, both are executed in the same mutate context.

File: builtins/amp-img.js (hunk 12 above)
Comment (jridgewell) [inline builtins/amp-img.js]: Ohh, ok.

**Comment by choumx** [comment]
Thanks for the reviews!

**Pull request merged**